Fix coding standards & introduce GitHub action.

peterwilsoncc opened the pull request:
<!-- Source of this Pull Request. Remove any that are not applicable. -->

**Issue**: #89

**Ticket**:

**Slack Thread**:

---

### Description

<!-- Describe the changes made in this Pull Request and the reason for these changes. -->

* Fixes all the coding standards warnings and errors for the plugin.
* Introduces a GitHub action for sniffing the coding standards.

Most of the missing `@since` annotations were for really old functions so I used the version number `1.4.13` as that was the earliest available tag in the check-in so I could be sure that the functions were included in that version.

Most of the changes are cosmetic but there are a few changes that might have side effects:

* changes to strict comparisons
* yoda-ing comparisons
* changes to recommended functions (eg `json_encode` to `wp_json_encode`)

I left [this line unchanged](https://github.com/woocommerce/woocommerce-gateway-payfast/pull/147/files#diff-2a669ad5948aeb08bee95f76bbd5384d566e54504a64f5c7b8176e042a28cd87R386) as modifying it requires a significant refactor that is probably best done as a dedicated PR to make sure nothing breaks.

I also left `urlencode` calls as is, I didn't want to risk the payfast API chocking on `%20` in place of `+`.

### Steps to Test

<!-- Describe the steps to replicate the issue and confirm the fix. -->
<!-- Include as many details as possible. -->

This changes much of the plugin, there are main class includes 513 changes (about 1/3rd of the file).  The following items should be checked:

* Purchases
   * Standard
   * Pre-order
   * Subscription
   * Subscription renewal
   * Subscription cancellation
   * Unsuccessful payment
* Privacy exports
   * Standard orders removed
   * Inactive subscriptions removed
   * Active subscription retained (needed for cancelling)

### Documentation

<!-- Will this change require new documentation or changes to existing documentation? -->
<!-- A good way to answer it is to ask: will more than one customer ever need to know about this? -->
- [ ] This PR needs documentation (has the "Documentation" label).
<!-- For an extra 💯 include further details about which change requires documentation. -->

### Changelog Entry

> Dev - Resolve coding standards issues.

Closes #89 .

peterwilsoncc committed:
Fix auto-fixable coding standard errors and warnings.

peterwilsoncc committed:
Exclude existing file from invalid name check.

peterwilsoncc committed:
Introduce/fix file comments.

peterwilsoncc committed:
Fix code sniffing errors.

peterwilsoncc committed:
Fix code sniffing warnings.

peterwilsoncc committed:
Coding standards: add docblocks.

peterwilsoncc committed:
Fix coding standards errors.

peterwilsoncc committed:
Add PHP coding standards tests.

peterwilsoncc committed:
Show sniffs in messages.

dkotter commented on .github/workflows/php-compatibility.yml:
```diff
@@ -28,4 +28,29 @@ jobs:
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Suggest we update this to v3

peterwilsoncc replied:
Updated in 9e59aff2f4d4145cb10618362a73149bad63708a.

dkotter commented on includes/class-wc-gateway-payfast.php:
```diff
@@ -1203,24 +1258,30 @@ public function cancel_subscription_listener( $subscription ) {
 	 * @return string
 	 */
-	protected function _generate_parameter_string( $api_data, $sort_data_before_merge = true, $skip_empty_values = true ) {
+	protected function _generate_parameter_string( $api_data, $sort_data_before_merge = true, $skip_empty_values = true ) {  // phpcs:ignore PSR2.Methods.MethodDeclaration.Underscore
```
Wondering if we should globally disable this rule since we're having to ignore that on multiple lines within this file?

peterwilsoncc replied:
> Wondering if we should globally disable this rule since we're having to ignore that on multiple lines within this file?

Done in aac4e4db25f3bb290f1ee55edfb983cff2e52b90. I excluded the rule for this file only as none of the other files were affected.

peterwilsoncc committed:
Exclude `PSR2.Methods.MethodDeclaration.Underscor` in affected file.

peterwilsoncc committed:
Exclude build folder from sniffs.

peterwilsoncc committed:
Update checkout action to v3.

vikrampm1 commented:
@ankitguptaindia This has been in QA status for a while now, so bumping this up in your queue.

ankitguptaindia commented:
Hello @peterwilsoncc Could you please fix the conflict in this PR?

peterwilsoncc committed:
Merge branch 'trunk' into fix/89-phpcs-sniffs

# Conflicts:
#	.github/workflows/php-compatibility.yml
#	gateway-payfast.php

peterwilsoncc commented:
@ankitguptaindia The merge conflicts are resolved.

ankitguptaindia commented:
Simple Product Purchase: 
- Order Updated. ✅ 
- ITN Updated  ✅ 

Subscription Purchase:
- Order Updated ✅ 
- ITN Updated  ✅ 

Renewal Subscription (from customer side):
- Order Updated ✅ 
- ITN Updated ✅ 

Subscription Cancellation:
- From Customer Side ✅ 

Subscription Reactivation and Payment (from customer side):

- Order Updated ✅ 
- ITN Updated ✅ 

Unsuccessful Payment:

- Order Updated ✅ 

Canceled Transactions: ✅ 


![Screenshot 2023-10-07 at 5 29 59 PM](https://github.com/woocommerce/woocommerce-gateway-payfast/assets/6384299/45fcdb52-6ab5-4eeb-864e-118c31dd7a1e)


----

QA Updates -  WIP 

Remains - Pre-order, Privacy exports (need to discuss with Dev)

peterwilsoncc commented:
@ankitguptaindia Following up your slack message re:privacy exports:

Firstly, I made a typo and meant _erasing_ private data, `wp-admin/erase-personal-data.php`.

The plugin ties in to the WordPress functionality for erasing data and will remove a customers orders when their email is entered on the screen above.

**Subscriptions**

* Orders containing active or on-hold subscriptions [are retained](https://github.com/woocommerce/woocommerce-gateway-payfast/blob/842bc37f95e7d4c9253f055e7b3dc1910cb7a10d/includes/class-wc-gateway-payfast-privacy.php#L246-L258) 
* Orders containing subscriptions with other status are removed, the payfast payment token is deleted first. [Source code](https://github.com/woocommerce/woocommerce-gateway-payfast/blob/842bc37f95e7d4c9253f055e7b3dc1910cb7a10d/includes/class-wc-gateway-payfast-privacy.php#L262-L270).

**Pre-orders**

* Pre-orders are deleted and the pre-order tokens are removed as part of this, this is to ensure the user is not charged when the pre-order release date arrives.

ankitguptaindia commented:
I tested this PR by following the provided instructions. I encountered some issues during the testing, but these issues are not specific to the changes made in this PR. I was able to reproduce these issues with older versions of the PayFast extension, including versions 1.5.0, 1.5.5, and 1.5.8. I have reported these issues separately for further discussion and potential fixes.

The issues reported are as follows:

- [Payment Fails when Pre Order Product has 'Upon Release' setting enabled for Payment](https://github.com/woocommerce/woocommerce-gateway-payfast/issues/162)

- [Renewal Orders Fail in Subscription Products with PayFast](https://github.com/woocommerce/woocommerce-gateway-payfast/issues/163) - Note that this issue was due to the Sandbox environment, and it has been closed.


The PR functions the same way as the stable version of this extension.

jeffpaul commented:
@faisal-alvi when we get back to the PHPCS work, please look at merging in or pulling across work from @peterwilsoncc's PR here into #160 so that we've got a single PR that handles these updates for PayFast... thanks!

faisal-alvi commented:
#160 is merged in the `trunk`, we can now merge the `trunk` here and proceed with the pending work.

jeffpaul commented:
@faisal-alvi would you please help take over this PR to handle that `trunk` merge & getting any additional/helpful changes from this PR ready for final review/merge?

faisal-alvi committed:
Merge branch 'trunk' into fix/89-phpcs-sniffs

faisal-alvi committed:
remove duplicate exclude-pattern

faisal-alvi committed:
fix phpcs for base url,
https://github.com/woocommerce/woocommerce-gateway-payfast/actions/runs/7178169594/job/19545867589#step:9:17
